magma/protocol_type.py

```diff
@@ -31,6 +31,9 @@ def _from_magma_value_(cls, val: Type):
     def flip(cls):
         return cls._flip_magma_()
 
+    def __len__(cls):
+        return len(cls._to_magma_())
+
```

Comment:
This was the simplest fix for getting the length of an arbitrary protocol, but it reserves the len method, an alternative is to update everywhere in the code where we call len(T) and replace it with len(T._to_magma_()) for protocols, this would avoid needing to reserve the method but add quite a bit of code changes, so I'll only do this if we anticipate the need for a custom len method

Reply:
Hmmm, this isn't ideal but will work for now.  I think in the long term if we want to have good support for protocols and allow magma to extend its interface we will need to use a sort of "dereference" pattern where before interacting with objects they are first unpacked. like is done in sequential: https://github.com/phanrahan/magma/blob/master/magma/syntax/sequential.py#L21

I realize to will be annoying to always need to call protocol methods before use.  However, it would allow magma to seamlessly extend its interfaces without requiring updates from protocol types.   It would also allow protocols to not have the mixin methods.  Finally it would remove the need to actually inherit the protocol and instead allow structural subtyping similar to the bitvector protocol (I could see this being especially import for the metaclass).  

https://github.com/leonardt/hwtypes/blob/master/tests/test_bv_protocol.py#L38

We could create a decorator which could ease some cases:
```Python
def unpack(value_or_type):
    if isinstance(value_or_type, MagmaProtocol):
        return value_or_type._get_magma_value_()
    elif isinstance(value_or_type, MagmaProtocolMeta):
        return value_or_type._to_magma_()
    else:
        return value_or_type

def cast_to_magma(fn):
    @wraps(fn)
    def wrapper(*args, **kwargs):
        args = (unpack(x) for x in args)
        kwargs = {k: unpack(v) for k, v in kwargs.items()}
        return fn(*args, **kwargs)
    return wrapper
```

Reply:
It should be noted I definitely can't add most `__dunders__` I don't currently use `__len__` but I do use others.

Reply:
Yea I agree that would be the most extensible, it's mainly a matter of laziness at this point ;) I think the proper long term solution is to have all these magma type level interactions go through the protocol (and have the protocol implemented by built-in types) so we won't need to do any of the "isinstance(..., protocol)" checks